File: 2024/typescript/day15.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

type Move = '^' | 'v' | '<' | '>'
type Field = '.' | '#' | 'O' | '@' | '[' | ']'
type Pos = [number, number]
type Grid = Field[][]
type State = {
    grid: Grid
    moves: Move[]
    robot: Pos
}

function parse(input: string): State {
    const [gridChunk, movesChunk] = utils.split_sections(input)
    const grid = utils
        .split_lines(gridChunk)
        .map(line => line.split('') as Field[])
    const moves = movesChunk.replaceAll('\n', '').split('') as Move[]
    const robot = grid
        .flatMap((row, y) => row.map((cell, x) => [cell, x, y]))
        .find(([cell]) => cell === '@')!
        .slice(1) as Pos
    return { grid, moves, robot }
}

function scaleUp({ grid, moves, robot: [x, y] }: State): State {
    grid = grid.map(row =>
        row
            .map(cell => {
                switch (cell) {
                    case '@':
                        return '@.'
                    case 'O':
                        return '[]'
                    case '.':
                        return '..'
                    case '#':
                        return '##'
                    default:
                        utils.unreachable(`Invalid field: ${cell}`)
                }
            })
            .join('')
            .split(''),
    ) as Grid
    return { grid, moves, robot: [x * 2, y] }
}

function nextPos([x, y]: Pos, move: Move): Pos {
    switch (move) {
        case '^':
            return [x, y - 1]
        case 'v':
            return [x, y + 1]
        case '<':
            return [x - 1, y]
        case '>':
            return [x + 1, y]
        default:
            utils.unreachable(`Invalid move: ${move}`)
    }
}

function canMove(grid: Grid, [x, y]: Pos, move: Move): boolean {
    const [nx, ny] = nextPos([x, y], move)
    switch (grid[ny][nx]) {
        case '.':
            return true
        case 'O':
            return canMove(grid, [nx, ny], move)
        case '[':
            switch (move) {
                case '<':
                    return canMove(grid, [nx, ny], move)
                case '>':
                    return canMove(grid, [nx + 1, ny], move)
                case '^':
                case 'v':
                    return (
                        canMove(grid, [nx, ny], move) &&
                        canMove(grid, [nx + 1, ny], move)
                    )
                default:
                    utils.unreachable(`Invalid move: ${move}`)
            }
        case ']':
            switch (move) {
                case '<':
                    return canMove(grid, [nx - 1, ny], move)
                case '>':
                    return canMove(grid, [nx, ny], move)
                case '^':
                case 'v':
                    return (
                        canMove(grid, [nx, ny], move) &&
                        canMove(grid, [nx - 1, ny], move)
                    )
                default:
                    utils.unreachable(`Invalid move: ${move}`)
            }
        case '#':
            return false
        default:
            utils.unreachable(`Invalid field: ${grid[ny][nx]}`)
    }
}

function doMove(grid: Grid, [x, y]: Pos, move: Move): Pos {
    const [nx, ny] = nextPos([x, y], move)

    switch (grid[ny][nx]) {
        case '.':
            break
        case 'O':
            doMove(grid, [nx, ny], move)
            break
        case '[':
            if (move === '>') {
                doMove(grid, [nx + 1, ny], move)
                doMove(grid, [nx, ny], move)
            } else {
                doMove(grid, [nx, ny], move)
                doMove(grid, [nx + 1, ny], move)
            }
            break
        case ']':
            if (move === '<') {
                doMove(grid, [nx - 1, ny], move)
                doMove(grid, [nx, ny], move)
            } else {
                doMove(grid, [nx, ny], move)
                doMove(grid, [nx - 1, ny], move)
            }
            break
        case '#':
            utils.unreachable('Invalid move: hit a wall')
        default:
            utils.unreachable(`Invalid field: ${grid[ny][nx]}`)
    }

    grid[ny][nx] = grid[y][x]
    grid[y][x] = '.'
    return [nx, ny]
}

function tryMove(grid: Grid, [x, y]: Pos, move: Move): [boolean, Pos] {
    const ok = canMove(grid, [x, y], move)
    return [ok, ok ? doMove(grid, [x, y], move) : [x, y]]
}

function moveRobot(state: State) {
    const { grid, moves, robot } = state
    const [_, pos] = moves.reduce(
        ([_, pos], move) => tryMove(grid, pos, move),
        [true, robot],
    )
    return { grid, moves: [], robot: pos }
}

function gpsSum(state: State): number {
    return state.grid
        .flatMap((row, y) =>
            row.map((cell, x) => [cell, x, y] as [Field, number, number]),
        )
        .filter(([cell]) => cell === 'O' || cell === '[')
        .reduce((acc, [_, x, y]) => acc + 100 * y + x, 0)
}

export const part1: Part = input => () => gpsSum(moveRobot(parse(input)))

export const part2: Part = input => () =>
    gpsSum(moveRobot(scaleUp(parse(input))))

export const day = import.meta.file.match(/day(\d+)/)![1]
export const input = await utils.readInputText(day)
part1.solution = 1563092
part2.solution = 1582688

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_text([
        `
########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

<^^>>>vv<v>>v<<
`,
        `
##########
#..O..O.O#
#......O.#
#.OO..O.O#
#..O@..O.#
#O#..O...#
#O..O..O.#
#.OO.O.OO#
#....O...#
##########

<vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
<<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
>^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
<><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^
`,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                ['Test part 1', 2028, part1(testInput[0])],
                ['Test part 1', 10092, part1(testInput[1])],
                ['Test part 2', 9021, part2(testInput[1])],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = number
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2024/zig/src/day15.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc_utils");

const Field = enum(u8) {
    wall = '#',
    empty = '.',
    box = 'O',
    box_left = '[',
    box_right = ']',
    robot = '@',
};
const Grid = aoc.Grid(Field, u8);

const Move = enum(u8) { up = '^', down = 'v', left = '<', right = '>', no_move = '\n' };

const Errors = error{InvalidInput};

const State = struct {
    grid: Grid,
    moves: []const Move,

    pub fn init(input: []const u8, gpa: Allocator) !State {
        var chunks = aoc.splitChunks(input);
        const grid: Grid = try .init(chunks.next() orelse return Errors.InvalidInput, gpa);
        const moves: []const Move = @ptrCast(chunks.next() orelse return Errors.InvalidInput);
        return .{ .grid = grid, .moves = moves };
    }

    pub fn deinit(self: *const State, gpa: Allocator) void {
        self.grid.deinit(gpa);
    }

    pub fn scaleUp(self: *State, gpa: Allocator) !void {
        var buf = try gpa.alloc(Field, self.grid.buf.len * 2);
        errdefer gpa.free(buf);
        for (self.grid.buf, 0..) |field, i| {
            const j = i * 2;
            const fields: [2]Field = switch (field) {
                .wall => .{ .wall, .wall },
                .empty => .{ .empty, .empty },
                .box => .{ .box_left, .box_right },
                .box_left => unreachable,
                .box_right => unreachable,
                .robot => .{ .robot, .empty },
            };
            @memcpy(buf[j .. j + 2], &fields);
        }
        gpa.free(self.grid.buf);
        self.grid.buf = buf;
        self.grid.width *= 2;
    }

    pub fn robot(self: State) Grid.Pos {
        return self.grid.findScalar(.robot).?;
    }

    fn nextPos(from: Grid.Pos, dir: Move) Grid.Pos {
        return switch (dir) {
            .up => .{ .x = from.x, .y = from.y - 1 },
            .down => .{ .x = from.x, .y = from.y + 1 },
            .left => .{ .x = from.x - 1, .y = from.y },
            .right => .{ .x = from.x + 1, .y = from.y },
            .no_move => unreachable,
        };
    }

    pub fn canMove(self: State, from: Grid.Pos, dir: Move) bool {
        const to = nextPos(from, dir);
        const field = self.grid.at(to);
        return switch (field) {
            .wall => false,
            .empty => true,
            .box => self.canMove(to, dir),
            .box_left => blk: {
                switch (dir) {
                    .right => break :blk self.canMove(nextPos(to, .right), dir),
                    .up, .down => break :blk self.canMove(to, dir) and self.canMove(nextPos(to, .right), dir),
                    .left, .no_move => unreachable,
                }
            },
            .box_right => blk: {
                switch (dir) {
                    .left => break :blk self.canMove(nextPos(to, .left), dir),
                    .up, .down => break :blk self.canMove(to, dir) and self.canMove(nextPos(to, .left), dir),
                    .right, .no_move => unreachable,
                }
            },
            .robot => unreachable,
        };
    }

    fn move(self: *State, from: Grid.Pos, to: Grid.Pos) void {
        const fromField = self.grid.at(from);
        self.grid.setAt(to, fromField);
        self.grid.setAt(from, .empty);
    }

    pub fn doMove(self: *State, from: Grid.Pos, dir: Move) Grid.Pos {
        const to = nextPos(from, dir);
        const toField = self.grid.at(to);
        switch (toField) {
            .wall => unreachable,
            .empty => {
                self.move(from, to);
            },
            .box => {
                _ = self.doMove(to, dir);
                self.move(from, to);
            },
            .box_left => {
                const nextTo = nextPos(to, .right);
                switch (dir) {
                    .right => {
                        _ = self.doMove(nextTo, dir);
                        self.grid.setAt(nextTo, toField);
                        self.move(from, to);
                    },
                    .up, .down => {
                        _ = self.doMove(to, dir);
                        _ = self.doMove(nextTo, dir);
                        self.move(from, to);
                    },
                    .left, .no_move => unreachable,
                }
            },
            .box_right => {
                const nextTo = nextPos(to, .left);
                switch (dir) {
                    .left => {
                        _ = self.doMove(nextTo, dir);
                        self.grid.setAt(nextTo, toField);
                        self.move(from, to);
                    },
                    .up, .down => {
                        _ = self.doMove(to, dir);
                        _ = self.doMove(nextTo, dir);
                        self.move(from, to);
                    },
                    .right, .no_move => unreachable,
                }
            },
            .robot => unreachable,
        }
        return to;
    }

    pub fn moveRobot(self: *State) void {
        var r = self.robot();
        for (self.moves) |m| {
            if (m != .no_move and self.canMove(r, m)) {
                r = self.doMove(r, m);
            }
        }
    }

    pub fn gpsCoordSum(self: State) u32 {
        var sum: u32 = 0;
        for (self.grid.buf, 0..) |field, i| {
            if (field == .box or field == .box_left) {
                const pos = self.grid.indexToPos(i);
                sum += @as(u32, pos.x) + 100 * @as(u32, pos.y);
            }
        }
        return sum;
    }
};

pub fn part1(input: []const u8, gpa: Allocator) !Day.Result1 {
    var state: State = try .init(input, gpa);
    defer state.deinit(gpa);
    state.moveRobot();
    return state.gpsCoordSum();
}

pub fn part2(input: []const u8, gpa: Allocator) !Day.Result2 {
    var state: State = try .init(input, gpa);
    defer state.deinit(gpa);
    try state.scaleUp(gpa);
    state.moveRobot();
    return state.gpsCoordSum();
}

pub const Day = aoc.DayInfo("15", u32, u32, 1563092, 1582688, @This(), &.{
    .{
        .expected1 = 2028,
        .expected2 = null,
        .input =
        \\########
        \\#..O.O.#
        \\##@.O..#
        \\#...O..#
        \\#.#.O..#
        \\#...O..#
        \\#......#
        \\########
        \\
        \\<^^>>>vv<v>>v<<
        ,
    },
    .{
        .expected1 = 10092,
        .expected2 = 9021,
        .input =
        \\##########
        \\#..O..O.O#
        \\#......O.#
        \\#.OO..O.O#
        \\#..O@..O.#
        \\#O#..O...#
        \\#O..O..O.#
        \\#.OO.O.OO#
        \\#....O...#
        \\##########
        \\
        \\<vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
        \\vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
        \\><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
        \\<<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
        \\^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
        \\^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
        \\>^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
        \\<><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
        \\^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
        \\v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^
        ,
    },
});

test "samples 1" {
    try Day.testPart1Samples();
}
test "samples 2" {
    try Day.testPart2Samples();
}
test "part 1" {
    try Day.testPart1();
}
test "part 2" {
    try Day.testPart2();
}
